### ISSUE TITLE
Autoscaling to zero

### ISSUE THREAD

**Issue by mkjpryor** [kubernetes-sigs/cluster-api-provider-openstack#1328]
/kind feature

**Describe the solution you'd like**
Auto-scaling to and from zero nodes has been implemented upstream - https://github.com/kubernetes/autoscaler/pull/4840.

We should make sure that we implement the provider side of this contract.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Comment by braghettos** [comment]
Adding a +1 because it would be a great feature!

**Comment by mdbooth** [comment]
https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20210310-opt-in-autoscaling-from-zero.md

**Comment by nikParasyr** [comment]
IIRC autoscaling from 0 already works, but it requires extra labels/annotations for the cluster-autoscaler. Implementing this would remove the need for these extra labels/annotations so it would be a bit easier for the user. 

For me, this wouldnt be very high prio

**Comment by mkjpryor** [comment]
> IIRC autoscaling from 0 already works, but it requires extra labels/annotations for the cluster-autoscaler. Implementing this would remove the need for these extra labels/annotations so it would be a bit easier for the user.
> 
> For me, this wouldnt be very high prio

This is correct I think. However I don't really want to have to query the flavor and add custom annotations to my machine deployment when a mechanism exists for the infrastructure provider to do this for me. Agree it isn't high, high priority but would certainly be good to implement the support since the hooks already exist in CAPI.

**Comment by mdbooth** [comment]
@nikParasyr @mkjpryor Do either of you want to work on this, btw? Seems like a good feature.

**Comment by nikParasyr** [comment]
@mdbooth I dont have any time to work on this (unfortunately)

**Comment by mkjpryor** [comment]
@mdbooth We think a customer might be about to ask for this. If they do then we will work on it.

**Comment by Ritikaa96** [comment]
This is still a active one , 
/reopen

**Comment by Ritikaa96** [comment]
Got it. 
@mkjpryor I think this is still a requirement and should be worked on.
Please do the needful in case it should be kept active.

**Comment by nikParasyr** [comment]
@lentzi90 could you please re-open this?

Im looking into implementing this ( coming weeks ) but before doing so, I'd like to run my approach through you to see if its valid:
- In general the contract for doing this is documented [here](https://cluster-api.sigs.k8s.io/developer/providers/contracts/infra-machine#inframachinetemplate-support-cluster-autoscaling-from-zero)
- Since it requires changing OpenStackMachineTemplate.Status we need a new controller for this. An implementation from a different provider is [here](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/main/controllers/vmware/vspheremachinetemplate_controller.go)
- I'm considering adding this controller and feature behind a feature gate the same way that cluster-api does for ClusterClass, ... ( see [1](https://github.com/kubernetes-sigs/cluster-api/tree/main/feature), [2](https://github.com/kubernetes-sigs/cluster-api/blob/main/main.go#L581) ): the reason for this is to be able to release this in alpha and also to limit the controllers users have when not needing this. 
- To fill the resource and nodeInfo im going to use attributes of the flavor and image ( there are "well_known" attributes for all the required info )
- For e2e testing: It should be relatively easy to add some tests on the clusterClass e2e tests validating that status fields are set. Should i also look to install the auto-scaler and test whether auto-scaling occurs? this will be more difficult 
- Lastly: I expect the PR will be rather big but still everything is related and is good to be viewed "together". I would do a single PR with multiple commits splitting functionality ( adding featureGate, controller, ... ). would that work for you or you prefer seperate PRs?

Thanks a lot

**Comment by lentzi90** [comment]
Thanks for bringing this up @nikParasyr !
The plan sounds good to me.

> For e2e testing: It should be relatively easy to add some tests on the clusterClass e2e tests validating that status fields are set. Should i also look to install the auto-scaler and test whether auto-scaling occurs? this will be more difficult

It would definitely be cool to have this properly tested. However, if it turns out to be too complicated I am fine leaving that out of the initial implementation. That said, I would expect that we can re-use [this CAPI e2e test](https://github.com/kubernetes-sigs/cluster-api/blob/main/test/e2e/autoscaler.go) quite easily.

> Lastly: I expect the PR will be rather big but still everything is related and is good to be viewed "together". I would do a single PR with multiple commits splitting functionality ( adding featureGate, controller, ... ). would that work for you or you prefer seperate PRs?

One PR with multiple commits sounds good for this kind of thing 🙂

**Comment by sd109** [comment]
I'm very interested in seeing this feature added too, so would be happy to help out with code / review / beta testing in the wild if any of that is useful.
